### PR TITLE
NCSU Canvas Rendering Project Initial Steps

### PR DESCRIPTION
**Pull request by Brody-Eastwood**
Implements the initial steps from:

https://github.com/servo/servo/wiki/Canvas-rendering-project

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

- [X] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/20447)

### CODE REVIEW

**Review by jdm — CHANGES_REQUESTED** [review]
Great work!

**Comment by jdm** [comment]
`./mach test-tidy` is complaining:
```
./components/layout/fragment.rs:11: missing space after ,
./components/script/dom/bindings/trace.rs:34: use statement is not in alphabetical order
	expected: canvas_traits::canvas::CanvasId
	found: canvas_traits::canvas::{CanvasGradientStop, LinearGradientStyle, RadialGradientStyle}
```
Please [squash](https://www.devroom.io/2011/07/05/git-squash-your-latests-commits-into-one/) all the commits into one as well.

**Comment by pmocher** [comment]
Hi! This is my first time squashing all of the commits into one. I probably did something that I wasn't supposed to.

**Comment by jdm** [comment]
Yeah, if the pull request still shows  >1 commit after refreshing, it didn't achieve the desired effect. Be sure to run `git rebase -i 5f62a1789a9bf1f4ddc736dd131d0fe4269242c3^` and mark all the commits as `s` instead of `pick` except for the topmost one.

**Comment by pmocher** [comment]
I see.  I'll do that now.

**Comment by pmocher** [comment]
It seems as though I've messed something up.

**Comment by jdm** [comment]
Indeed. I recommend running `git reflog`, copying the hash of the entry that corresponds to the last commit you made that appears in that list of git actions, and running `git reset --hard [hash]`. This should reset the state of your repository to the pre-squash state.

**Comment by jdm** [comment]
Somehow you keep including merges with master in the squash, I think :(

**Comment by Brody-Eastwood** [comment]
I tried several times to get things squashed. This is the only time I've gotten it to say that it's only one commit, but given how much this pull request says it changes, I think I failed again?

**Comment by Brody-Eastwood** [comment]
I've tried several times to get it to squash. This is the only time I've gotten it to actually only be one commit, but given how much this pull request says it changes (and the fact that it ceased to be able to be automerged) I think I failed again.

**Comment by jdm** [comment]
If you reset the branch to its former state with lots of commits, I'll squash it and write down what I did.

**Comment by Brody-Eastwood** [comment]
Thank you!

I've reverted it.

Sorry for being so difficult.

**Comment by jdm** [comment]
It's ok; this is an ongoing problem with people making their first contribution to Servo and we haven't figured out a good way to deal with it yet.

**Comment by jdm** [comment]
For future reference, I recommend following https://github.com/servo/servo/wiki/Github-workflow.

**Comment by jdm** [comment]
Since the branch includes merges, I decided it would be easier to cherry-pick the actual commits we care about into a new branch:
```
# my master branch is pristine from upstream; I don’t make any changes to it locally.
$ git checkout master
$ git checkout -b squashed
$ git cherry-pick 5f62a17
$ git cherry-pick 6a3b013
# resolve merge conflict that is reported by git status
$ git add components/script/dom/canvasrenderingcontext2d.rs
$ git cherry-pick --continue
$ git cherry-pick 28d4ce82dc133502d605f722afeadf0591ae755d
# ignore the fact that the commit is marked as empty
$ git cherry-pick 492302b349c6ceafea558c433ae1d6e0eed90f82
# ignore the fact that the commit is marked as empty
$ git cherry-pick 8dee649e6707c2d23dd8098fea40fca1c7331b50
# resolve merge conflict reported by git status
$ git add components/canvas/canvas_paint_thread.rs
$ git cherry-pick --continue
$ git cherry-pick f4024e76b31a16f6a74983925e6324622551d78f
# resolve merge conflict reported by git status
$ git add components/script/dom/htmlcanvaselement.rs components/script/dom/canvasrenderingcontext2d.rs
$ git cherry-pick --continue
$ git cherry-pick 2bcf001e232979751dd51d15178826cb9ae6cc80
$ git cherry-pick 125bfbd00ca10b2c1a5ce2e1225168f6bb1c935f
$ git cherry-pick 0016441cb1a5b57ce4f7f480b7726ee86d128d8b
$ git cherry-pick 9b5a353bd4488711189152eb4f83f75766efaefd
$ git cherry-pick d048508fc8cc296bbe52c0f4f94292a9b66b15df
$ git cherry-pick 46cbe7b95b1a9398f928417ed5f19a6e5826401d
$ git cherry-pick 33bb495f2d6d2c2acd6aed68b9fcfabffd836321
# resolve merge conflict reported by git status
$ git add components/canvas/canvas_paint_thread.rs components/script/dom/canvasrenderingcontext2d.rs components/script/dom/htmlcanvaselement.rs
$ git cherry-pick --continue
# Now we can squash the result.
$ git rebase -i 76d548c99b676dfc0c12c9d87e03269e6ee55adc^

—
pick 76d548c99b initial shot at initial steps
f 61b360e8f4 Second shot at initial steps
f 6606486d72 Progress on initial steps
f 3666516bd5 Initial steps finally builds!
f f7a92f7b3e timed test to drawImage from one canvas to another
f e8fcc50fdd setting up canvas id in layout
f 6730fde6a5 finished associating CanvasId to layout
f 2626b4169c removed 'unused import' build warning
f 96087b0618 Removed files that shouldn't be in the repo.
f d67797d7b0 Converted tabs to spaces
f 1885589ea0 first try at squashing
—

$ git commit --amend --no-edit -m "Add unique canvas IDs to all canvas operations."
$ git push -f Brody-Eastwood squashedcanvas:master
```

**Comment by jdm** [comment]
Ok, all of the automated tests that use a canvas are panicking with stacks like this:
```
  ▶ CRASH [expected PASS] /2dcontext/line-styles/canvas_linestyles_linecap_001.htm
  │ 
  │ VMware, Inc.
  │ softpipe
  │ 3.3 (Core Profile) Mesa 17.3.0-devel
  │ assertion failed: thread_state::get().is_script() (thread LayoutThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(NonZero(NonZero(1))) }, at components/script/dom/bindings/cell.rs:98)
  │ stack backtrace:
  │    0:     0x7f418c1b2d6c - backtrace::backtrace::trace::heb9efa3021143bc0
  │    1:     0x7f418c1b20f2 - backtrace::capture::Backtrace::new::hd20e533c9c999e36
  │    2:     0x7f41891ae90c - servo::main::{{closure}}::h96f2fd035380f7d9
  │    3:     0x7f418c1c1905 - std::panicking::rust_panic_with_hook::h489260a27605eec3
  │                         at libstd/panicking.rs:577
  │    4:     0x7f418a2c1fc3 - std::panicking::begin_panic::h88970b163c1446a9
  │    5:     0x7f4189b2a7dd - script::dom::htmlcanvaselement::HTMLCanvasElement::get_canvas_id::h8af515f893db25df
  │    6:     0x7f4189824166 - <script::dom::bindings::root::LayoutDom<script::dom::htmlcanvaselement::HTMLCanvasElement> as script::dom::htmlcanvaselement::LayoutHTMLCanvasElementHelpers>::data::hb960f9960efcb3dc
  │    7:     0x7f418982de02 - <script::dom::bindings::root::LayoutDom<script::dom::node::Node> as script::dom::node::LayoutNodeHelpers>::canvas_data::hc4d0763ec1019568
  │    8:     0x7f41895358fe - <layout::construct::FlowConstructor<'a, ConcreteThreadSafeLayoutNode>>::build_fragment_for_block::h313458c542f016e8
  │    9:     0x7f41895234bf - <layout::construct::FlowConstructor<'a, ConcreteThreadSafeLayoutNode> as layout::traversal::PostorderNodeMutTraversal<ConcreteThreadSafeLayoutNode>>::process::ha6229ff0a27676fd
  │   10:     0x7f4189630eef - style::traversal::DomTraversal::handle_postorder_traversal::h992f91b2749e6bd8
  │   11:     0x7f41895ca3dc - style::driver::traverse_dom::h9b6af5485b4e1784
  │   12:     0x7f41894da9d0 - layout_thread::LayoutThread::handle_reflow::hafe580cf2b79e1af
  │   13:     0x7f41894d447c - layout_thread::LayoutThread::handle_request_helper::hb3abd95d5d0b09c0
  │   14:     0x7f41894d2d79 - layout_thread::LayoutThread::start::h9537b41eb8b948cf
  │   15:     0x7f41895ba063 - profile_traits::mem::ProfilerChan::run_with_memory_reporting::hc7bbe6eff5174296
  │   16:     0x7f418953f62b - std::sys_common::backtrace::__rust_begin_short_backtrace::h7bf083075dc64c41
  │   17:     0x7f41894b54c7 - _ZN3std9panicking3try7do_call17hdeaf1406e5993f92E.llvm.9F59020
  │   18:     0x7f418c1eaf9e - __rust_maybe_catch_panic
  │                         at libpanic_unwind/lib.rs:102
  │   19:     0x7f418959f23a - <F as alloc::boxed::FnBox<A>>::call_box::hdb8301d823bd613e
  │   20:     0x7f418c1d98a7 - <alloc::boxed::Box<alloc::boxed::FnBox<A, Output$u3d$R$GT$$u20$$u2b$$u20$$u27$a$GT$$u20$as$u20$core..ops..function..FnOnce$LT$A$GT$$GT$::call_once::h4bbca053e9002218
  │                         at /checkout/src/liballoc/boxed.rs:798
  │                          - std::sys_common::thread::start_thread::h3301b0b4cb89e2c1
  │                         at libstd/sys_common/thread.rs:24
  │   21:     0x7f418c1c1f68 - std::sys::unix::thread::Thread::new::thread_start::h6f2358c4012c0e12
  │                         at libstd/sys/unix/thread.rs:90
  │   22:     0x7f418809f183 - start_thread
  │   23:     0x7f4186966ffc - clone
  │   24:                0x0 - <unknown>
```

This means that LayoutHTMLCanvasElementHelpers::data is invoking HTMLCanvasElement::get_canvas_id, which calls DomRefCell::borrow on `self.context`. This fails assertions we have in place about which thread is active when this borrow operation occurs, so you'll need to write a `get_canvas_id_for_layout` helper method that calls duplicates `get_canvas_id` but uses `borrow_for_layout` instead of `borrow`. Sorry, it's not a great system.

**Comment by jdm** [comment]
cc @SimonSapin so you're aware of what's happening in this PR.

**Comment by jdm** [comment]
Note: to run these 2d and 3d canvas tests locally, use `./mach test-wpt tests/wpt/web-platform-tests/2dcontext/transformations/canvas_transformations_reset_001.html` and `./mach test-wpt tests/wpt/mozilla/tests/webgl/conformance-2.0.0/conformance2/samplers/sampler-drawing-test.html`. This will verify that any changes made actually allow the tests to run correctly.

File: components/script/dom/htmlcanvaselement.rs [hunk 1]
```diff
@@ -156,6 +158,18 @@ impl LayoutHTMLCanvasElementHelpers for LayoutDom<HTMLCanvasElement> {
                 .unwrap_or(LengthOrPercentageOrAuto::Auto)
         }
     }
+
+    #[allow(unsafe_code)]
+    fn get_canvas_id_for_layout(&self) -> CanvasId {
+        unsafe {
+            let canvas = &*self.unsafe_get();
+            if let &Some(CanvasContext::Context2d(ref context)) = canvas.context.borrow_for_layout() {
+                context.get_canvas_id()
```
Comment (jdm) [inline components/script/dom/htmlcanvaselement.rs]: Sorry, one more change is required. There needs to be a `get_canvas_id_for_layout` method added to `LayoutCanvasRenderingContext2DHelpers` in `canvasrenderingcontext2d.rs`. This line should then read `context.to_layout().get_canvas_id()`.

**Review by jdm — CHANGES_REQUESTED** [review]

**Comment by Brody-Eastwood** [comment]
Those tests succeeded locally! I'm really hoping we'll see success fully this time!

**Pull request merged**